File: src/main/java/com/traffic/backend/service/FineService.java
package com.traffic.backend.service;

import com.traffic.backend.dto.FineDTO;
import com.traffic.backend.model.Fine;
import com.traffic.backend.model.User;
import com.traffic.backend.repository.FineRepository;
import com.traffic.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class FineService {

    private final FineRepository fineRepository;
    private final UserRepository userRepository;

    public Fine issueFine(FineDTO dto) {
        User user = userRepository.findById(dto.getUserId())
                .orElseThrow(() -> new RuntimeException("User not found"));

        Fine fine = new Fine();
        fine.setVehiclePlate(dto.getVehiclePlate());
        fine.setReason(dto.getReason());
        fine.setAmount(dto.getAmount());
        fine.setStatus("UNPAID");
        fine.setIssuedAt(LocalDateTime.now());
        fine.setUser(user);

        return fineRepository.save(fine);
    }

    public List<Fine> getAllFines() {
        return fineRepository.findAll();
    }

    public Fine getFineById(Long id) {
        return fineRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Fine not found"));
    }

    public List<Fine> getFinesByUser(Long userId) {
        return fineRepository.findByUserId(userId);
    }

    public List<Fine> getFinesByVehicle(String plate) {
        return fineRepository.findByVehiclePlate(plate);
    }

    public Fine updateFine(Long id, FineDTO dto) {
        Fine fine = getFineById(id);
        fine.setVehiclePlate(dto.getVehiclePlate());
        fine.setReason(dto.getReason());
        fine.setAmount(dto.getAmount());
        return fineRepository.save(fine);
    }

    public Fine markAsPaid(Long id) {
        Fine fine = getFineById(id);
        fine.setStatus("PAID");
        return fineRepository.save(fine);
    }

    public void deleteFine(Long id) {
        fineRepository.deleteById(id);
    }
}

File: web-user/traffic-fine-management/src/services/fineService.js
import api from "./api";

export async function getAllFines() {
  const res = await api.get("/api/fines");
  return res.data;
}

export async function getFineById(id) {
  const res = await api.get(`/api/fines/${id}`);
  return res.data;
}

export async function getFinesByUser(userId) {
  const res = await api.get(`/api/fines/user/${userId}`);
  return res.data;
}

export async function getFinesByVehicle(plate) {
  const res = await api.get(`/api/fines/vehicle/${encodeURIComponent(plate)}`);
  return res.data;
}

export async function issueFine(fineDto) {
  const res = await api.post("/api/fines", fineDto);
  return res.data;
}

export async function updateFine(id, fineDto) {
  const res = await api.put(`/api/fines/${id}`, fineDto);
  return res.data;
}

export async function markFineAsPaid(id) {
  const res = await api.put(`/api/fines/${id}/pay`);
  return res.data;
}

export async function deleteFine(id) {
  const res = await api.delete(`/api/fines/${id}`);
  return res.data;
}

export default {
  getAllFines,
  getFineById,
  getFinesByUser,
  getFinesByVehicle,
  issueFine,
  updateFine,
  markFineAsPaid,
  deleteFine,
};
